src/test/java/com/uid2/admin/managers/KeysetManagerTest.java

```diff
@@ -0,0 +1,133 @@
+package com.uid2.admin.managers;
+
+import com.uid2.admin.secret.IKeysetKeyManager;
+import com.uid2.admin.store.writer.KeysetStoreWriter;
+import com.uid2.admin.util.KeysetUtil;
+import com.uid2.shared.auth.ClientKey;
+import com.uid2.shared.auth.Keyset;
+import com.uid2.shared.auth.KeysetSnapshot;
+import com.uid2.shared.auth.Role;
+import com.uid2.shared.store.reader.RotatingKeysetProvider;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentMatcher;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.Map;
+
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.ArgumentMatchers.isNull;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+public class KeysetManagerTest {
+    protected AutoCloseable mocks;
+
+    @Mock protected KeysetStoreWriter keysetStoreWriter;
+    @Mock protected RotatingKeysetProvider keysetProvider;
+    @Mock
+    protected IKeysetKeyManager keysetKeyManager;
+    @Mock protected KeysetSnapshot keysetSnapshot;
+
+    @BeforeEach
+    public void setupMocks() {
+        mocks = MockitoAnnotations.openMocks(this);
+        when(keysetProvider.getSnapshot()).thenReturn(keysetSnapshot);
+    }
+
+    @AfterEach
+    public void teardown() throws Exception {
+        mocks.close();
+    }
+
+    protected void setKeysets(Map<Integer, Keyset> keysets) {
+        when(keysetSnapshot.getAllKeysets()).thenReturn(keysets);
+    }
+
+    @Test
+    public void testAddOrReplaceKeyset() throws Exception{
+        KeysetManager keysetManager = new KeysetManager(keysetProvider, keysetStoreWriter,
+                keysetKeyManager, true);
+
+        Map<Integer, Keyset> keysets = new HashMap<Integer, Keyset>() {{
+                put(1, KeysetUtil.createDefaultKeyset(3, 1));
+                put(2, KeysetUtil.createDefaultKeyset(4, 2));
+                put(3, KeysetUtil.createDefaultKeyset(5, 3));
+                put(4, KeysetUtil.createDefaultKeyset(6, 4));
+            }};
+
+        setKeysets(keysets);
+
+        // add new keyset
+        Keyset keyset1 = KeysetUtil.createDefaultKeyset(7, 5);
+        keysetManager.addOrReplaceKeyset(keyset1);
+        assertTrue(keysets.containsKey(5));
+        assertTrue(keyset1.equals(keysets.get(5)));
+        verify(keysetStoreWriter).upload(mapOfSize(5), isNull());
```

Comment:
should we test that the actual new keyset was uploaded

Reply:
The line above checks that

Reply:
kinda - we check map size for upload() but not for keysets, but then we check correct keysets are in the snapshot but not whether they're uploaded. It's probably fine - just a little inconsistent

Reply:
Do you have an example?

Reply:
no, cost/benefit probably not worth fixing this